core/matrix/dense.cpp

```diff
@@ -1032,20 +1050,134 @@ std::unique_ptr<Dense<ValueType>> Dense<ValueType>::row_gather(
     return result;
 }
 
+template <typename ValueType>
+void Dense<ValueType>::row_gather(const Array<int32>* row_indices,
+                                  Dense<ValueType>* row_collection) const
+{
+    this->row_gather_impl(row_indices, row_collection);
+}
+
+template <typename ValueType>
+void Dense<ValueType>::row_gather(const Array<int64>* row_indices,
+                                  Dense<ValueType>* row_collection) const
+{
+    this->row_gather_impl(row_indices, row_collection);
+}
+
+
+namespace {
+
+
+/**
+ * run uses template to go through the list and select the valid
+ * template and run it.
+ *
+ * @tparam T  the type of input object
+ * @tparam Func  the function will run if the object can be converted to K
+ * @tparam ...Args  the additional arguments for the Func
+ *
+ * @note this is the end case
+ */
+template <typename T, typename Func, typename... Args>
+void run(T, Func, Args...)
+{
+    GKO_NOT_IMPLEMENTED;
+}
+
+/**
+ * run uses template to go through the list and select the valid
+ * template and run it.
+ *
+ * @tparam K  the current type tried in the convertion
+ * @tparam ...Types  the other types will be tried in the conversion if K fails
+ * @tparam T  the type of input object
+ * @tparam Func  the function will run if the object can be converted to K
+ * @tparam ...Args  the additional arguments for the Func
+ *
+ * @param obj  the input object waiting converted
+ * @param f  the function will run if obj can be converted successfully
+ * @param args  the additional arguments for the function
+ */
+template <typename K, typename... Types, typename T, typename Func,
+          typename... Args>
+void run(T obj, Func f, Args... args)
+{
+    if (auto dobj = dynamic_cast<K>(obj)) {
+        f(dobj, args...);
+    } else {
+        run<Types...>(obj, f, args...);
+    }
+}
```

Comment:
This should be moved to some utility header, we will most likely need it elsewhere again.